### PR TITLE
Test classification spans when testing signature help items

### LINKED ISSUE #50114: Visual Studio throw error when using a generic type having xml document containing `<list>` tag.
**Version Used**: 

Microsoft Visual Studio Community 2019
Version 16.8.3

**Steps to Reproduce**:

Define a generic type having xml document containing `<list>` tag.

```C#
/// <summary>Here is an example of a bulleted list:
/// <list type="bullet">
/// <item>
/// <description>Item 1.</description>
/// </item>
/// <item>
/// <description>Item 2.</description>
/// </item>
/// </list>
/// </summary>
public class Generic<T>{}
```

Use the type

```C#
/// <summary>Here is an example of a bulleted list:
/// <list type="bullet">
/// <item>
/// <description>Item 1.</description>
/// </item>
/// <item>
/// <description>Item 2.</description>
/// </item>
/// </list>
/// </summary>
public class Generic<T>{}

class Program
{
    static void Main(string[] args)
    {
        _ = new Generic<> // Visual Studio throws error when typing `<`
    }
}
```

**Expected Behavior**:

No error.

**Actual Behavior**:

Visual Studio throws error.

```
type        : Error
source      : Editor or Editor Extension
description : System.InvalidOperationException: Unexpected value 'ContainerStart' of type 'System.String'
                 場所 Microsoft.CodeAnalysis.TaggedTextExtensions.ToClassificationTypeName(String taggedTextTag)
                 場所 Microsoft.CodeAnalysis.Editor.Shared.Extensions.ClassificationExtensions.ToClassificationSpans(IEnumerable`1 parts, ITextSnapshot textSnapshot, ClassificationTypeMap typeMap)
                 場所 Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.SignatureHelp.Presentation.SignatureHelpClassifier.GetClassificationSpans(SnapshotSpan span)
                 場所 Microsoft.VisualStudio.Text.Classification.Implementation.ClassifierTagger.<GetTags>d__5.MoveNext()
                 場所 Microsoft.VisualStudio.Text.Tagging.Implementation.TagAggregator`1.<GetTagsForBuffer>d__47.MoveNext()
```

### PR DESCRIPTION
**Pull request by davidwengier**
Fixes #50114

If there are existing tests for this, I couldn't find them, so this is a little odd, so open to suggestions for a better approach.

### CODE REVIEW

File: src/EditorFeatures/CSharpTest/SignatureHelp/GenericNameSignatureHelpProviderTests.cs [hunk 1]
```diff
@@ -849,5 +850,44 @@ void Goo()
             var expectedOrderedItems = new List<SignatureHelpTestItem>();
             await TestAsync(markup, expectedOrderedItems);
         }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.SignatureHelp)]
+        public async Task DeclaringGenericTypeWithDocCommentList()
```
Comment (Youssef1313) [inline src/EditorFeatures/CSharpTest/SignatureHelp/GenericNameSignatureHelpProviderTests.cs:855]: Consider adding a WorkItem for the linked issue.

Comment (davidwengier) [reply src/EditorFeatures/CSharpTest/SignatureHelp/GenericNameSignatureHelpProviderTests.cs:855]: Good point, thanks!

File: src/EditorFeatures/TestUtilities/SignatureHelp/SignatureHelpTestItem.cs [hunk 2]
```diff
@@ -34,27 +34,33 @@ public struct SignatureHelpTestItem
...
         /// <summary>
         /// Whether this item is expected to be selected.
         /// Note: If no item is expected to be selected, the verification of the actual selected item is skipped.
         /// </summary>
         public readonly bool IsSelected;
 
+        /// <summary>
+        /// The classification spans for the method documentation
+        /// </summary>
+        public readonly IEnumerable<string>? ClassificationTypeNames;
```
Comment (sharwell) [inline src/EditorFeatures/TestUtilities/SignatureHelp/SignatureHelpTestItem.cs:53]:
```suggestion
        public readonly ImmutableArray<string>? ClassificationTypeNames;
```

File: src/EditorFeatures/TestUtilities/SignatureHelp/SignatureHelpTestItem.cs [hunk 3]
```diff
@@ -34,27 +34,33 @@ public struct SignatureHelpTestItem
...
+            string? methodDocumentation = null,
+            string? parameterDocumentation = null,
             int? currentParameterIndex = null,
-            string description = null,
-            string prettyPrintedSignature = null,
-            bool isSelected = false)
+            string? description = null,
+            string? prettyPrintedSignature = null,
+            bool isSelected = false,
+            IEnumerable<string>? classificationTypeNames = null)
```
Comment (sharwell) [inline src/EditorFeatures/TestUtilities/SignatureHelp/SignatureHelpTestItem.cs:63]:
```suggestion
            ImmutableArray<string>? classificationTypeNames = null)
```

File: src/EditorFeatures/TestUtilities/SignatureHelp/AbstractSignatureHelpProviderTests.cs [hunk 4]
```diff
@@ -242,6 +244,14 @@ private async Task VerifyCurrentParameterNameWorkerAsync(string markup, string e
             {
                 Assert.Equal(expectedTestItem.Description, ToString(actualSignatureHelpItem.DescriptionParts));
             }
+
+            // Always get and realise the classified spans, even if no expected spans are passed in, to at least validate that
+            // exceptions aren't thrown
+            var classifiedSpans = actualSignatureHelpItem.DocumentationFactory(CancellationToken.None).ToClassifiedSpans().ToList();
+            if (expectedTestItem.ClassificationTypeNames != null)
+            {
+                Assert.Equal(string.Join(", ", expectedTestItem.ClassificationTypeNames), string.Join(", ", classifiedSpans.Select(s => s.ClassificationType)));
```
Comment (sharwell) [inline src/EditorFeatures/TestUtilities/SignatureHelp/AbstractSignatureHelpProviderTests.cs:253]:
```suggestion
            if (expectedTestItem.ClassificationTypeNames is { } classificationTypeNames)
            {
                Assert.Equal(string.Join(", ", classificationTypeNames), string.Join(", ", classifiedSpans.Select(s => s.ClassificationType)));
```

File: src/EditorFeatures/CSharpTest/SignatureHelp/GenericNameSignatureHelpProviderTests.cs [hunk 5]
```diff
@@ -849,5 +850,45 @@ void Goo()
...
+            expectedOrderedItems.Add(new SignatureHelpTestItem("G<S, T>", "List:\r\n\r\nItem 1.",
+                classificationTypeNames: new List<string>
+                {
+                    ClassificationTypeNames.Text,
+                    ClassificationTypeNames.WhiteSpace,
+                    ClassificationTypeNames.WhiteSpace,
+                    ClassificationTypeNames.WhiteSpace,
+                    ClassificationTypeNames.Text,
+                    ClassificationTypeNames.WhiteSpace
+                }));
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/SignatureHelp/GenericNameSignatureHelpProviderTests.cs:889]:
```suggestion
                classificationTypeNames: ImmutableArray.Create(
                    ClassificationTypeNames.Text,
                    ClassificationTypeNames.WhiteSpace,
                    ClassificationTypeNames.WhiteSpace,
                    ClassificationTypeNames.WhiteSpace,
                    ClassificationTypeNames.Text,
                    ClassificationTypeNames.WhiteSpace)));
```

**Review by sharwell — APPROVED** [review]

**Pull request merged**